---
title: "R Notebook"
output: html_notebook
---



```{r}
#install remotes and BiocManager if necessary 
#if (!requireNamespace("remotes", quietly = TRUE))
#    install.packages("remotes")
#if (!requireNamespace("BiocManager", quietly = TRUE))
#    install.packages("BiocManager")

# additionally install several Bioconductor dependencies
# BiocManager::install(c("BSgenome.Hsapiens.UCSC.hg38", "EnsDb.Hsapiens.v86", 
                       "rtracklayer","GenomicFeatures","plyranges"))

# remotes::install_github("satijalab/PASTA")
# BiocManager::install("ComplexHeatmap")
```

```{r include=FALSE}
library(PASTA)
library(GenomicRanges)
library(dplyr)
library(ggplot2)
library(tibble)
library(ggthemes)
library(ComplexHeatmap)
library(circlize)
```

```{r}
hek <- readRDS("../../../../data/prmt-apa/CPA_HEK293FT.Rds")
hek
hek[["polyA"]][[]]
hek[["polyA"]]$counts
```

First task is to see if we can match up our PolyA sites of interest, with the sites in the Perturb-Seq data
1) how many are exact matches
2) perhaps I need to make 10-20nt windows to match them
So I need to get the PolyA sites we're interested in into the same format as the row IDs of the metadata table, or put everything into bed format. Putting everything into bed format is probably a better idea incase I need to adjust the ranges.


1. Our sites of interest
```{r}
# Define the folder path
folder_path <- "../../recount/bed/sites"

# Get the list of files in the folder
files <- list.files(path = folder_path, pattern = "\\.bed$", full.names = TRUE,recursive=FALSE)

# Filter for *not* the upstream bed regions
files <- files[!grepl("JTE|CFIM",files)]

df_list <- list()

# Loop through each file
for (file in files) {
  # Read the file into a dataframe
  df <- read.table(file, header = FALSE, sep = "\t", stringsAsFactors = FALSE)
  
  # Extract the filename without the path
  file_name <- basename(file)
  
  # Add the "condition" column
  df$condition <- sapply(file_name, function(file_name) {
    if (grepl("non_DMAi", file_name)) {
      case <- "non_DMAi"
   #  } else if (grepl("CFIM25", file_name)) {
  #     case <- "siCFIM25"
    } else if (grepl("control", file_name)) {
      case <- "control"
    } else if (grepl("mitigation", file_name)) {
      case <- "mitigation"  
    } else if (grepl("JTE607", file_name)) {
      case <- "JTE607"
    } else {
      case <- "DMAi"
    }
    
    switch(case,
      "non_DMAi" = "nonDMAi",
      "DMAi" = "DMAi",
      "control" = "control",
      "JTE607" = "JTE607",
    #   "siCFIM25" = "siCFIM25",
      "mitigation" = paste0("siCFIM25",strsplit(basename(file_name), "_")[[1]][2])
    )
  })

  # Add the "site_type" column
  if (grepl("distal", file_name)) {
    df$site_type <- "distal"
  } else if (grepl("proximal", file_name)) {
    df$site_type <- "proximal"
  }
  
  # Convert the dataframe to a GRanges object
  gr <- GRanges(seqnames = Rle(df$V1),
                ranges = IRanges(start = df$V2, end = df$V3),
                strand = Rle(df$V6),
                name = df$V4,
                condition = df$condition,
                site_type = df$site_type)
  # These are single nucleotide ranges, lets make them 20nt ranges with 10nt either side for matching with the other dataset ( i tried 4nt range and 10nt range - 4->10 gives you more, 10->20 is a couple more but not that many)
  gr <- resize(gr, width = 50, fix = "center")
  df_list[[file_name]] <- gr
}

length(df_list)
df_list
```

2. now to address the quantified perturb-seq sites
```{r}
pseq = hek[["polyA"]][[]]
pseq = pseq %>%
  filter(!is.na(hg38_Chromosome))
pseq_gr = GRanges(seqnames = Rle(pseq$hg38_Chromosome),
                ranges = IRanges(start = pseq$hg38_Position, end = pseq$hg38_Position_end),
                strand = Rle(pseq$strand),
                name = pseq$Ensemble_ID,
                condition = "",
                site_type = "")
pseq_gr
```
3. if a perturb-seq site overlaps with our site then label it with the condition and site type 
```{r}
for (gr in df_list) {
  # Find overlaps between pseq_gr and current gr
  overlaps <- findOverlaps(pseq_gr, gr, ignore.strand=FALSE)
  
  # Annotate pseq_gr with condition and site_type from the overlapping gr
  pseq_gr$condition[queryHits(overlaps)] <- gr$condition[subjectHits(overlaps)]
  pseq_gr$site_type[queryHits(overlaps)] <- gr$site_type[subjectHits(overlaps)]
}

matched_sites = pseq_gr[pseq_gr$condition != ""]

matched_sites
```

4. How many matched sites do we have in each category? How does this compare to original number?

```{r}

original_site_count = data.frame(
  name = names(df_list),
  number_original_sites = sapply(df_list, length)
) %>% filter(grepl("proximal",name))

original_site_count
original_site_count$condition <- sapply(original_site_count$name, function(file_name) {
    if (grepl("non_DMAi", file_name)) {
      case <- "non_DMAi"
   #  } else if (grepl("CFIM25", file_name)) {
  #     case <- "siCFIM25"
    } else if (grepl("control", file_name)) {
      case <- "control"
    } else if (grepl("mitigation", file_name)) {
      case <- "mitigation"  
    } else if (grepl("JTE607", file_name)) {
      case <- "JTE607"
    } else {
      case <- "DMAi"
    }
    
    switch(case,
      "non_DMAi" = "nonDMAi",
      "DMAi" = "DMAi",
      "control" = "control",
      "JTE607" = "JTE607",
    #   "siCFIM25" = "siCFIM25",
      "mitigation" = paste0("siCFIM25",strsplit(basename(file_name), "_")[[1]][2])
    )
  })



genes_with_both <- as.data.frame(matched_sites) %>%
  group_by(name) %>% arrange(name) %>%
  filter(n() == 2 & any(site_type == "proximal") & any(site_type == "distal")) %>%
  ungroup()

genes_with_both_summary <- as.data.frame(matched_sites) %>%
  group_by(name) %>% arrange(name) %>%
  filter(n() == 2 & any(site_type == "proximal") & any(site_type == "distal")) %>%
  ungroup() %>%
  group_by(condition, site_type) %>%
  summarise(count_found_in_CPAperturb = n()) %>%
  ungroup() %>%
  filter(grepl("proximal",site_type))


left_join(original_site_count, genes_with_both_summary, by=("condition"="condition")) %>%
  mutate(percent_found_in_CPAperturb=(count_found_in_CPAperturb/number_original_sites)*100) %>%
  select(condition, percent_found_in_CPAperturb, count_found_in_CPAperturb) %>%
  fwrite(.,file="percent-found-in-cpa-perturb-hek-DMAi-50nt-window.csv")
```
Ought to do the differential polyA analysis first, then filter for our sites, as I realise the outputs of this aren't provided with the original paper/in the seurat object I think.

```{r}
# WARNING this takes ages
# hek = CalcPolyAResiduals(hek, assay="polyA", background = "NT", gene.names = "Ensemble_ID")
```
```{r}
kd_list = Idents(hek) %>% unique() %>% as.character(.) 
kd_list = kd_list[kd_list != "NT"]

# all_res = list()
# for (protein in kd_list){
#   print(paste0("doing ",protein))
#   all_res[[protein]] = FindDifferentialPolyA(hek, ident.1 = protein, ident.2 = "NT",gene.names="Ensemble_ID")
# }
# saveRDS(all_res,file="../Outputs/CPA-Perturb-Seq-Differential.RData")
```

```{r}
all_res = readRDS("../../Outputs/CPA-Perturb-Seq-Differential.RData")
```


```{r}
pseq2 <- pseq %>%
  rownames_to_column("pseq_ID")

merged_ref_df <- merge(
  genes_with_both,
  pseq2,
  by.x = c("seqnames", "start", "end", "strand"),
  by.y = c("hg38_Chromosome", "hg38_Position", "hg38_Position_end", "strand")
)

merged_ref_df

for (protein in kd_list){
  left_join(all_res[[protein]] %>% rownames_to_column("pseq_ID"),merged_ref_df, by=c("pseq_ID"="pseq_ID")) %>% filter(!is.na(seqnames)) %>%
    ggplot(.,aes(x=condition,y=Estimate)) + facet_wrap(~site_type) + geom_boxplot() + theme_few() + ggtitle(protein)
  ggsave(paste0("../../Outputs/cpa-perturb-seq-",protein,".png"))
}
  
```

Make a big summary heatmap?

```{r}
reslist = list()
for (protein in kd_list){
  reslist[[protein]] = left_join(all_res[[protein]] %>% rownames_to_column("pseq_ID"),merged_ref_df, by=c("pseq_ID"="pseq_ID")) %>% filter(!is.na(seqnames)) %>%
    mutate(kd=protein)
}

res = do.call(rbind,reslist)

res_dist = res %>% filter(site_type=="distal")

# Ensure all conditions are covered
conditions <- unique(res_dist$condition)

# Define a color palette for the conditions
condition_colors <- c("darkseagreen4","darkslategray3","grey")
names(condition_colors) <- conditions

# Cap the Estimate values at 1 and -1
res_dist$Estimate <- pmax(pmin(res_dist$Estimate, 0.5), -0.5)

# Reshape data into a matrix form suitable for heatmap
heatmap_data <- reshape2::dcast(res_dist, pseq_ID ~ kd, value.var = "Estimate")
rownames(heatmap_data) <- heatmap_data$pseq_ID
heatmap_data <- as.matrix(heatmap_data[,-1])

# Define colors
heatmap_colors <- colorRamp2(c(min(heatmap_data, na.rm = TRUE), 0, max(heatmap_data, na.rm = TRUE)), c("blue", "white", "red"))

# Create a row annotation for conditions
row_anno <- rowAnnotation(
  condition = res_dist$condition[match(rownames(heatmap_data), res_dist$pseq_ID)],
  col = list(condition = condition_colors),
  annotation_legend_param = list(condition = list(title = "Condition"))
)

# Create the heatmap
ht <- Heatmap(
  heatmap_data,
  name = "Estimate",
  col = heatmap_colors,
  cluster_columns = TRUE,
  cluster_rows = TRUE,
  show_row_names = FALSE,
  show_column_names = TRUE,
  row_names_side = "left",
  row_split = res_dist$condition[match(rownames(heatmap_data), res_dist$pseq_ID)], # Split rows by condition
  gap = unit(2, "mm"),
  left_annotation = row_anno, # Add row annotation here
  heatmap_legend_param = list(
    title = "Estimate",
    title_position = "topcenter",
    legend_direction = "horizontal",
    legend_width = unit(4, "cm")
  )
)

pdf("../../Outputs/cpa-perturbseq-DMAI-heatmap.pdf")
# Draw the heatmap
draw(ht, heatmap_legend_side = "bottom", annotation_legend_side = "right")
dev.off()

```
```{r}

# 1. Build and combine your results, then filter to distal sites
reslist <- lapply(kd_list, function(protein) {
  all_res[[protein]] %>%
    rownames_to_column("pseq_ID") %>%
    left_join(merged_ref_df, by = "pseq_ID") %>%
    filter(!is.na(seqnames)) %>%
    mutate(kd = protein)
})
res <- bind_rows(reslist)
res_dist <- res %>% filter(site_type == "distal")

# 2. Define color palette (same as before)
conditions <- unique(res_dist$condition)
condition_colors <- setNames(
  c("darkseagreen4","darkslategray3","grey"),
  conditions
)

# 3. Cap Estimate values between –0.5 and +0.5
res_dist <- res_dist %>%
  mutate(Estimate = pmax(pmin(Estimate, 0.5), -0.5))

# 4. Order proteins by median Estimate in the “cfim full” condition
order_levels <- res_dist %>%
  filter(condition == "DMAi") %>%
  group_by(kd) %>%
  summarise(median_est = median(Estimate, na.rm = TRUE)) %>%
  arrange(median_est) %>%
  pull(kd)

res_dist <- res_dist %>%
  mutate(kd = factor(kd, levels = order_levels),
         condition = factor(condition, levels = c("DMAi","nonDMAi","control")))

# 5. Make grouped boxplots with proteins ordered by cfim full median
p <- ggplot(res_dist, aes(x = kd, y = Estimate, fill = condition)) +
  geom_boxplot(position = position_dodge(width = 0.8), outlier.size = 0.7) +
  scale_fill_manual(values = condition_colors) +
  labs(
    x = "CPA-PerturbSeq Knockdown",
    y = "Estimate",
    fill = "Condition"
  ) +
  theme_bw(base_size = 12) +
  theme(
    axis.text.x = element_text(angle = 45, hjust = 1),
    panel.grid.major.x = element_blank()
  )

# 6. Save to PDF
pdf("../../Outputs/dmai-cpa-perturbseq-boxplots-ordered-distal.pdf", width = 8, height = 5)
print(p)
dev.off()
```


Filtered by significant differences
```{r}

res = do.call(rbind,reslist)

res_dist = res %>% filter(site_type=="distal")
res_dist = res_dist %>% mutate(Estimate= ifelse(p_val_adj < 0.05,Estimate,0))

# Ensure all conditions are covered
conditions <- unique(res_dist$condition)

# Define a color palette for the conditions
condition_colors <- c("darkseagreen4","darkslategray3","grey")
names(condition_colors) <- conditions

# Cap the Estimate values at 1 and -1
res_dist$Estimate <- pmax(pmin(res_dist$Estimate, 0.5), -0.5)

# Reshape data into a matrix form suitable for heatmap
heatmap_data <- reshape2::dcast(res_dist, pseq_ID ~ kd, value.var = "Estimate")
rownames(heatmap_data) <- heatmap_data$pseq_ID
heatmap_data <- as.matrix(heatmap_data[,-1])

# Define colors
heatmap_colors <- colorRamp2(c(min(heatmap_data, na.rm = TRUE), 0, max(heatmap_data, na.rm = TRUE)), c("blue", "white", "red"))

# Create a row annotation for conditions
row_anno <- rowAnnotation(
  condition = res_dist$condition[match(rownames(heatmap_data), res_dist$pseq_ID)],
  col = list(condition = condition_colors),
  annotation_legend_param = list(condition = list(title = "Condition"))
)

# Create the heatmap
ht <- Heatmap(
  heatmap_data,
  name = "Estimate",
  col = heatmap_colors,
  cluster_columns = TRUE,
  cluster_rows = TRUE,
  show_row_names = FALSE,
  show_column_names = TRUE,
  row_names_side = "left",
  row_split = res_dist$condition[match(rownames(heatmap_data), res_dist$pseq_ID)], # Split rows by condition
  gap = unit(2, "mm"),
  left_annotation = row_anno, # Add row annotation here
  heatmap_legend_param = list(
    title = "Estimate",
    title_position = "topcenter",
    legend_direction = "horizontal",
    legend_width = unit(4, "cm")
  )
)

pdf("cpa-perturbseq-heatmap-sigFilt.pdf")
# Draw the heatmap
draw(ht, heatmap_legend_side = "bottom", annotation_legend_side = "right")
dev.off()
```
No capping and filter sig diff
```{r}
res = do.call(rbind,reslist)

res_dist = res %>% filter(site_type=="distal")
res_dist = res_dist %>% mutate(Estimate= ifelse(p_val_adj < 0.05,Estimate,0))

# Ensure all conditions are covered
conditions <- unique(res_dist$condition)

# Define a color palette for the conditions
condition_colors <- c("darkseagreen4","darkslategray3","grey")
names(condition_colors) <- conditions

# Cap the Estimate values at 1 and -1
# res_dist$Estimate <- pmax(pmin(res_dist$Estimate, 0.5), -0.5)

# Reshape data into a matrix form suitable for heatmap
heatmap_data <- reshape2::dcast(res_dist, pseq_ID ~ kd, value.var = "Estimate")
rownames(heatmap_data) <- heatmap_data$pseq_ID
heatmap_data <- as.matrix(heatmap_data[,-1])

# Define colors
heatmap_colors <- colorRamp2(c(min(heatmap_data, na.rm = TRUE), 0, max(heatmap_data, na.rm = TRUE)), c("blue", "white", "red"))

# Create a row annotation for conditions
row_anno <- rowAnnotation(
  condition = res_dist$condition[match(rownames(heatmap_data), res_dist$pseq_ID)],
  col = list(condition = condition_colors),
  annotation_legend_param = list(condition = list(title = "Condition"))
)

# Create the heatmap
ht <- Heatmap(
  heatmap_data,
  name = "Estimate",
  col = heatmap_colors,
  cluster_columns = TRUE,
  cluster_rows = TRUE,
  show_row_names = FALSE,
  show_column_names = TRUE,
  row_names_side = "left",
  row_split = res_dist$condition[match(rownames(heatmap_data), res_dist$pseq_ID)], # Split rows by condition
  gap = unit(2, "mm"),
  left_annotation = row_anno, # Add row annotation here
  heatmap_legend_param = list(
    title = "Estimate",
    title_position = "topcenter",
    legend_direction = "horizontal",
    legend_width = unit(4, "cm")
  )
)

pdf("cpa-perturbseq-heatmap-sigFilt-nocap.pdf")
# Draw the heatmap
draw(ht, heatmap_legend_side = "bottom", annotation_legend_side = "right")
dev.off()
```

```{r}
res = do.call(rbind,reslist)
res_dist = res %>% filter(site_type=="distal")

# Calculate the summary statistics
summary_data <- res_dist %>%
  group_by(kd, condition) %>%
  summarise(
    total = n(),
    total_change = sum(p_val_adj < 0.05 & abs(Estimate) >= 0.2),
    fraction_changed = total_change / total
  ) %>%
  ungroup()

# Filter the summary data for DMAi condition to reorder the kd factor levels
order_data <- summary_data %>%
  filter(condition == "DMAi") %>%
  arrange(desc(fraction_changed))

# Reorder the kd factor levels
summary_data <- summary_data %>%
  mutate(kd = factor(kd, levels = order_data$kd))

# Create the bar chart
ggplot(summary_data, aes(x = kd, y = fraction_changed, fill = condition)) +
  geom_bar(position = "dodge", stat = "identity") +
  scale_fill_manual(values = condition_colors) +
  theme_few() +
  theme(axis.text.x = element_text(angle = 90, hjust = 1)) +
  labs(x = "KD", y = "Fraction Changed", title = "Fraction Changed by KD and Condition")
ggsave("fraction-changed.png")
```

