---
title: "R Notebook"
output: html_notebook
---



```{r}
#install remotes and BiocManager if necessary 
#if (!requireNamespace("remotes", quietly = TRUE))
#    install.packages("remotes")
#if (!requireNamespace("BiocManager", quietly = TRUE))
#    install.packages("BiocManager")

#additionally install several Bioconductor dependencies
#BiocManager::install(c("BSgenome.Hsapiens.UCSC.hg38", "EnsDb.Hsapiens.v86", 
                       "rtracklayer","GenomicFeatures","plyranges"))

#remotes::install_github("satijalab/PASTA")
```

```{r}
library(PASTA)
library(GenomicRanges)
library(dplyr)
library(ggplot2)
library(tibble)
library(ggthemes)
```

```{r}
hek <- readRDS("../../../data/prmt-apa/CPA_HEK293FT.Rds")
hek
hek[["polyA"]][[]]
hek[["polyA"]]$counts
```

First task is to see if we can match up our PolyA sites of interest, with the sites in the Perturb-Seq data
1) how many are exact matches
2) perhaps I need to make 10-20nt windows to match them
So I need to get the PolyA sites we're interested in into the same format as the row IDs of the metadata table, or put everything into bed format. Putting everything into bed format is probably a better idea incase I need to adjust the ranges.


1. Our sites of interest
```{r}
# Define the folder path
folder_path <- "../motif_and_clip_analysis/DMAi_vs_non_DMAi/bed"

# Get the list of files in the folder
files <- list.files(path = folder_path, pattern = "\\.bed$", full.names = TRUE)

df_list <- list()

# Loop through each file
for (file in files) {
  # Read the file into a dataframe
  df <- read.table(file, header = FALSE, sep = "\t", stringsAsFactors = FALSE)
  
  # Extract the filename without the path
  file_name <- basename(file)
  
  # Add the "condition" column
  if (grepl("DMAi", file_name)) {
    if (grepl("non_DMAi", file_name)) {
      df$condition <- "nonDMAi"
    } else {
      df$condition <- "DMAi"
    }
  } else if (grepl("control", file_name)) {
    df$condition <- "control"
  }
  
  # Add the "site_type" column
  if (grepl("distal", file_name)) {
    df$site_type <- "distal"
  } else if (grepl("proximal", file_name)) {
    df$site_type <- "proximal"
  }
  
  # Convert the dataframe to a GRanges object
  gr <- GRanges(seqnames = Rle(df$V1),
                ranges = IRanges(start = df$V2, end = df$V3),
                strand = Rle(df$V6),
                name = df$V4,
                condition = df$condition,
                site_type = df$site_type)
  # These are 400nt ranges with 200nt flanking the polyA site on either side, lets make them 20nt ranges with 10nt either side for matching with the other dataset ( i tried 4nt range and 10nt range - 4->10 gives you more, 10->20 is a couple more but not that many)
  gr <- resize(gr, width = 20, fix = "center")
  df_list[[file_name]] <- gr
}

df_list
```

2. now to address the quantified perturb-seq sites
```{r}
pseq = hek[["polyA"]][[]]
pseq = pseq %>%
  filter(!is.na(hg38_Chromosome))
pseq_gr = GRanges(seqnames = Rle(pseq$hg38_Chromosome),
                ranges = IRanges(start = pseq$hg38_Position, end = pseq$hg38_Position_end),
                strand = Rle(pseq$strand),
                name = pseq$Ensemble_ID,
                condition = "",
                site_type = "")
pseq_gr
```
3. if a perturb-seq site overlaps with our site then label it with the condition and site type 
```{r}
for (gr in df_list) {
  # Find overlaps between pseq_gr and current gr
  overlaps <- findOverlaps(pseq_gr, gr, ignore.strand=FALSE)
  
  # Annotate pseq_gr with condition and site_type from the overlapping gr
  pseq_gr$condition[queryHits(overlaps)] <- gr$condition[subjectHits(overlaps)]
  pseq_gr$site_type[queryHits(overlaps)] <- gr$site_type[subjectHits(overlaps)]
}

matched_sites = pseq_gr[pseq_gr$condition != ""]
```

How many matched sites do we have in each category?

```{r}
df = as.data.frame(matched_sites)
df %>%
  group_by(condition, site_type) %>%
  summarise(count = n())

# check the matched proximal and distal sites

genes_with_both <- df %>%
  group_by(name) %>%
  filter(n() == 2 & any(site_type == "proximal") & any(site_type == "distal")) %>%
  ungroup()

genes_with_both %>%
  group_by(condition, site_type) %>%
  summarise(count = n())
```
Ought to do the differential polyA analysis first, then filter for our sites, as I realise the outputs of this aren't provided with the original paper/in the seurat object I think.

```{r}
# WARNING this takes ages
hek = CalcPolyAResiduals(hek, assay="polyA", background = "NT", gene.names = "Ensemble_ID")
```
```{r}
# WARNING this also takes ages
kd_list = Idents(hek) %>% unique() %>% as.character(.) 
kd_list = kd_list[kd_list != "NT"]

all_res = list()
for (protein in kd_list){
  print(paste0("doing ",protein))
  all_res[[protein]] = FindDifferentialPolyA(hek, ident.1 = protein, ident.2 = "NT",gene.names="Ensemble_ID")
}
saveRDS(all_res,file="../Outputs/CPA-Perturb-Seq-Differential.RData")
```

```{r}
pseq2 <- pseq %>%
  rownames_to_column("pseq_ID")

merged_ref_df <- merge(
  genes_with_both,
  pseq2,
  by.x = c("seqnames", "start", "end", "strand"),
  by.y = c("hg38_Chromosome", "hg38_Position", "hg38_Position_end", "strand")
)

merged_ref_df

for (protein in kd_list){
  left_join(all_res[[protein]] %>% rownames_to_column("pseq_ID"),merged_ref_df, by=c("pseq_ID"="pseq_ID")) %>% filter(!is.na(seqnames)) %>%
    ggplot(.,aes(x=condition,y=Estimate)) + facet_wrap(~site_type) + geom_boxplot() + theme_few() + ggtitle(protein)
  ggsave(paste0("../Outputs/cpa-perturb-seq-",protein,".png"))
}
  
```

Make a big summary heatmap?

```{r}
for (protein in kd_list){
  left_join(all_res[[protein]] %>% rownames_to_column("pseq_ID"),merged_ref_df, by=c("pseq_ID"="pseq_ID")) %>% filter(!is.na(seqnames)) %>%
    mutate(kd=protein)
}
```

